### DISCUSSION TITLE
defined?(yield) vs block_given?

### DISCUSSION THREAD

**Discussion by davekaro** [jeremyevans/sequel#2149] [General]
Starting in [5.49.0](https://github.com/jeremyevans/sequel/blob/4314aebb3bbc4f1aa4b32083b42b10e5818aad8f/doc/release_notes/5.49.0.txt#L54) sequel switched to using "defined?(yield) instead of block_given? internally for better performance on CRuby". I ran some recent benchmarks and found `block_given?` to be faster. I realized the reason was I had YJIT enabled, and in https://github.com/ruby/ruby/pull/7202 `block_given?` became YJIT optimized. How do you think about YJIT with respect to performance optimizations like this? Should sequel change again to take advantage of YJIT, but end up being slower if you don't use it? Something else?

**Comment by jeremyevans** [comment]
Are you testing on Ruby 3.3?  Looks like YJIT didn't start optimizing `defined?(yield)` until after the Ruby 3.3 release. See ruby/ruby@3c9290173a1421b0624a6d62c0844c778dbc61ad.  I assume with current ruby master, `defined?(yield)` is faster.  It should always be faster, because it doesn't have to check whether the method has been overridden.

**Reply by davekaro** [reply]
Yes Ruby 3.3. Here is my script:

```
require "benchmark"
require "benchmark/ips"

RubyVM::YJIT.enable

def defined_yield
  yield if defined?(yield)
end

def block_given
  yield if block_given?
end

Benchmark.ips do |x|
  x.report("defined?(yield)") {
    defined_yield
  }

  x.report("block_given?") {
    block_given
  }

  x.compare!
end
```

and the results:

```
ruby bench.rb
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) +YJIT [arm64-darwin23]
Warming up --------------------------------------
     defined?(yield)     3.268M i/100ms
        block_given?     4.010M i/100ms
Calculating -------------------------------------
     defined?(yield)     59.950M (± 0.2%) i/s -    300.674M in   5.015402s
        block_given?     64.555M (± 1.1%) i/s -    324.842M in   5.032632s

Comparison:
        block_given?: 64555304.3 i/s
     defined?(yield): 59950475.8 i/s - 1.08x  slower
```

If I remove the `RubyVM::YJIT.enable` this is my result

```
ruby bench.rb
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [arm64-darwin23]
Warming up --------------------------------------
     defined?(yield)     3.138M i/100ms
        block_given?     2.090M i/100ms
Calculating -------------------------------------
     defined?(yield)     31.553M (± 1.2%) i/s -    160.013M in   5.072006s
        block_given?     21.130M (± 2.5%) i/s -    106.611M in   5.048559s

Comparison:
     defined?(yield): 31552684.4 i/s
        block_given?: 21130127.9 i/s - 1.49x  slower
```

Looks like Ruby 3.3 is optimizing it.

**Reply by davekaro** [reply]
Sorry @jeremyevans after I typed that all out, realized you said `defined?(yield)` wasn't optimized yet! Your reasoning does make sense to me. Thanks!
